File: tests/test_mysql.rs
#[cfg(test)]
mod test_mysql {

    use serde_json::Value::{Array, Null, Number, String};
    #[macro_export(local_inner_macros)]
    macro_rules! mysql {
        ($($json:tt)+) => {{
            let expr = jsql::query::decode::decode_express(&serde_json::json!($($json)+));
            jsql::dialects::mysql::mysql(expr.unwrap())
        }}
    }
    #[test]
    fn test_eq_omit() {
        let (sql, params) = mysql!({"name": "apple"});
        assert_eq!(sql, "`name` = ?");
        assert_eq!(params, vec![String("apple".into())])
    }

    #[test]
    fn test_eq() {
        let (sql, params) = mysql!({"name":{"$eq": "apple"}});
        assert_eq!(sql, "`name` = ?");
        assert_eq!(params, vec![String("apple".into())])
    }

    #[test]
    fn test_ne() {
        let (sql, params) = mysql!({"name": {"$ne":"apple"}});
        assert_eq!(sql, "`name` != ?");
        assert_eq!(params, vec![String("apple".into())])
    }
    #[test]
    fn test_gt() {
        let (sql, params) = mysql!({"weight": {"$gt":10}});
        assert_eq!(sql, "`weight` > ?");
        assert_eq!(params, vec![Number(10.into())])
    }
    #[test]
    fn test_lt() {
        let (sql, params) = mysql!({"weight": {"$lt":30}});
        assert_eq!(sql, "`weight` < ?");
        assert_eq!(params, vec![Number(30.into())])
    }

    #[test]
    fn test_gt_lt() {
        let (sql, params) = mysql!({"weight": {"$gt":10, "$lt":30}});
        assert_eq!(sql, "`weight` > ? and `weight` < ?");
        assert_eq!(params, vec![Number(10.into()), Number(30.into())])
    }

    #[test]
    fn test_and_omit() {
        let (sql, params) = mysql!({"name": "apple", "weight": {"$gt":10, "$lt":30}});
        assert_eq!(sql, "`name` = ? and `weight` > ? and `weight` < ?");
        assert_eq!(
            params,
            vec![String("apple".into()), Number(10.into()), Number(30.into())]
        )
    }
    #[test]
    fn test_and() {
        let (sql, params) = mysql!({"$and":[{"name": "apple"},{"weight": {"$gt":10, "$lt":30}}]});
        assert_eq!(sql, "`name` = ? and `weight` > ? and `weight` < ?");
        assert_eq!(
            params,
            vec![String("apple".into()), Number(10.into()), Number(30.into())]
        )
    }

    #[test]
    fn test_or() {
        let (sql, params) = mysql!({"$or": [{"name": "apple"}, {"name": "orange"}]});
        assert_eq!(sql, "`name` = ? or `name` = ?");
        assert_eq!(
            params,
            vec![String("apple".into()), String("orange".into())]
        )
    }

    #[test]
    fn test_mmysql() {
        let (sql, params) = mysql!({"name": "apple", "weight": {"$gt":10, "$lt":30}});
        assert_eq!(sql, "`name` = ? and `weight` > ? and `weight` < ?");
        assert_eq!(
            params,
            vec![String("apple".into()), Number(10.into()), Number(30.into())]
        )
    }

    #[test]
    fn test_and_unpack() {
        let (sql, params) =
            mysql!({"name": "apple", "$and":[{"weight": {"$gt":10}},{"weight": {"$lt":30}}]});
        assert_eq!(sql, "`weight` > ? and `weight` < ? and `name` = ?");
        assert_eq!(
            params,
            vec![Number(10.into()), Number(30.into()), String("apple".into()),]
        );

        let (sql, params) = mysql!({"$and":[{"$and":[{"weight": {"$gt":10}},{"weight": {"$lt":30}}]},{"name": "apple"}]});
        assert_eq!(sql, "`weight` > ? and `weight` < ? and `name` = ?");
        assert_eq!(
            params,
            vec![Number(10.into()), Number(30.into()), String("apple".into()),]
        )
    }

    #[test]
    fn test_null() {
        let (sql, params) = mysql!({"name":{"$eq": null}});
        assert_eq!(sql, "`name` = ?");
        assert_eq!(params, vec![Null])
    }

    #[test]
    fn test_in() {
        let (sql, params) = mysql!({"name":{"$in": ["apple", "banana"]}});
        assert_eq!(sql, "`name` in ?");
        assert_eq!(params.len(), 1);
        assert_eq!(
            params[0],
            Array(vec![String("apple".into()), String("banana".into())])
        );
    }

    #[test]
    fn test_nin() {
        let (sql, params) = mysql!({"name":{"$nin": ["apple", "banana"]}});
        assert_eq!(sql, "`name` not in ?");
        assert_eq!(params.len(), 1);
        assert_eq!(
            params[0],
            Array(vec![String("apple".into()), String("banana".into())])
        );
    }

    #[test]
    fn test_regex() {
        let (sql, params) = mysql!({"name":{"$regex": "app?"}});
        assert_eq!(sql, "`name` REGEXP ?");
        assert_eq!(params, vec![String("app?".into())]);
    }

    // #[test]
    // fn test_mysql() {
    //     let json = json!({ "name": "apple", "qty": {"$gt":10, "$lt": 30 }, "$and": [{"color":"red"}, {"status": "ok"}], "$or": [{"level": "A"},{"level": "B"},{"level": "C"}] });
    //     let express = decode_express(json);
    //     assert_eq!(mysql(express), "`color` = ? and `status` = ? and ( `level` = ? or `level` = ? or `level` = ?)  and `name` = ? and `qty` > ? and `qty` < ?");
    // }
}

File: tests/test_mysql.py
from datetime import datetime

import jsql


def test_input_bytes():
    sql, params = jsql.mysql('{"name":{"$eq": "apple"}}'.encode("utf-8"))
    assert sql == "`name` = ?"
    assert params == ("apple",)


def test_input_str():
    sql, params = jsql.mysql('{"name":{"$eq": "apple"}}')
    assert sql == "`name` = ?"
    assert params == ("apple",)


def test_input_dict():
    sql, params = jsql.mysql({"name": {"$eq": "apple"}})
    assert sql == "`name` = ?"
    assert params == ("apple",)


def test_eq():
    sql, params = jsql.mysql('{"name":{"$eq": "apple"}}')
    assert sql == "`name` = ?"
    assert params == ("apple",)

    sql, params = jsql.mysql('{"name":"apple"}')
    assert sql == "`name` = ?"
    assert params == ("apple",)


def test_in():
    sql, params = jsql.mysql({"size": {"$in": [1, 2, 3]}})
    assert sql == "`size` in ?"
    assert params == ([1, 2, 3],)


def test_gt_lt():
    sql, params = jsql.mysql({"date": {"$gt": datetime(2024, 2, 1), "$lt": datetime(2024, 2, 10)}})
    assert sql == "`date` > ? and `date` < ?"
    assert params == (
        datetime(2024, 2, 1),
        datetime(2024, 2, 10),
    )


def test_mysql():
    input = {
        "name": "apple",
        "date": {"$gt": datetime(2024, 2, 1), "$lt": datetime(2024, 2, 10)},
        "$or": [{"colour": "red"}, {"colour": "green"}],
    }
    sql, params = jsql.mysql(input)
    assert sql == "`name` = ? and `date` > ? and `date` < ? and (`colour` = ? or `colour` = ?)", sql
    assert params == ("apple", datetime(2024, 2, 1, 0, 0), datetime(2024, 2, 10, 0, 0), "red", "green")



# TODO: add test cases
